# Import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/jagpreetbatth/Desktop/Career Foundry/Python/11-2021 Instacart Basket Analysis'



In [3]:
path

'/Users/jagpreetbatth/Desktop/Career Foundry/Python/11-2021 Instacart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_prods.shape

(49693, 5)

In [7]:
df_ords.shape

(3421083, 7)

In [8]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [10]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# Mixed Type Data - Create Data frame

In [11]:
df_test = pd.DataFrame()

In [12]:
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# Check for Mixed Types

In [15]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


# Change data type of mix column to string

In [16]:
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
df_test['mix'].dtype

dtype('O')

# Missing Values

In [18]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

# Find the Missing Values

In [19]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [20]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
df_prods.shape

(49693, 5)

# Create new dataframe that excludes the missing values

In [23]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [24]:
 df_prods_clean.shape

(49677, 5)

# Duplicates

In [25]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
df_prods_clean.shape

(49677, 5)

# Create New Data Frame without Duplicates 

In [29]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
df_prods_clean_no_dups.shape

(49672, 5)

In [31]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'products_checked.csv'))

# Data Consitency for Products Table (continued)

In [32]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

## Note the 'prices' column has a very high maximum when compared to the 75% (3rd quartile). Error in entering data or actual high price possibility. 

# Data Consitency for Orders Table

# Check for Mixed Types

In [33]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## No 'mix' out = no mixed data type columns found

# Check for Missing Values

In [34]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## Note the 'days_since_prior_order' column has 206209 missing values. Due to lack of reporting, unavailable data of prior customer orders or first time customers who never had previous orders.

## Fix the Missing Values Found

In [44]:
df_ords.shape

(3421083, 7)

## Create New Data Frame without Missing Values

In [40]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [42]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   
5           5   3367565        1             6                   2   

   order_hour_of_day  days_since_prior_order  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0  
5                  7                    19.0

## Shape check to see if 206209 Missing values were removed

In [43]:
df_ords_clean.shape

(3214874, 7)

## Removed Missing Values by creating a new dataframe to exclude them as they do not effect the overall data and comprise a very small percentage of the whole data set.

# Check for Duplicates

## Create Subset Limited to Duplicates

In [45]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [46]:
df_ords_dups.head()

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## No duplicates found

## Remove 'Unnamed:0' column

In [47]:
df_ords_clean_no_dups = df_ords_clean.drop(columns = ['Unnamed: 0'])

In [48]:
df_ords_clean_no_dups.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   

   days_since_prior_order  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0

# Export clean products data

In [49]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_checked.csv'))

# Export clean orders data

In [50]:
df_ords_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_checked.csv'))